firewood/src/proof.rs

```diff
@@ -455,93 +447,51 @@ impl<N: AsRef<[u8]> + Send> Proof<N> {
         buf: &[u8],
         end_node: bool,
     ) -> Result<(DiskAddress, Option<SubProof>, usize), ProofError> {
-        let mut items: Vec<Encoded<Vec<u8>>> = bincode::DefaultOptions::new().deserialize(buf)?;
-        let size = items.len();
-
-        match size {
-            EXT_NODE_SIZE => {
-                let mut items = items.into_iter();
-
-                let cur_key_path: Vec<u8> = items
-                    .next()
-                    .unwrap()
-                    .decode()?
-                    .into_iter()
-                    .flat_map(to_nibble_array)
-                    .collect();
-
-                let (cur_key_path, term) = PartialPath::decode(&cur_key_path);
-                let cur_key = cur_key_path.into_inner();
-
-                let data: Vec<u8> = items.next().unwrap().decode()?;
-
+        let node = NodeType::decode(buf)?;
+        let addr = merkle
+            .new_node(Node::new(node.clone()))
+            .map(|node| node.as_ptr())
+            .map_err(ProofError::InvalidNode)?;
+        match node {
```

Comment:
Avoid the clone here:
```suggestion
        let new_node = merkle
            .new_node(Node::new(node))
            .map_err(ProofError::InvalidNode)?;
        let addr = new_node.as_ptr();
        match new_node.inner() {
```